src/library/trackset/setlogfeature.cpp

```diff
@@ -447,69 +579,122 @@ void SetlogFeature::slotPlayingTrackChanged(TrackPointer currentPlayingTrack) {
     }
 }
 
-void SetlogFeature::slotPlaylistContentChanged(QSet<int> playlistIds) {
+void SetlogFeature::slotPlaylistContentOrLockChanged(const QSet<int>& playlistIds) {
+    // qDebug() << "slotPlaylistContentOrLockChanged() playlistId:" << playlistId;
+    QSet<int> idsToBeUpdated;
     for (const auto playlistId : qAsConst(playlistIds)) {
-        enum PlaylistDAO::HiddenType type =
-                m_playlistDao.getHiddenType(playlistId);
-        if (type == PlaylistDAO::PLHT_SET_LOG ||
-                type == PlaylistDAO::PLHT_UNKNOWN) { // In case of a deleted Playlist
-            updateChildModel(playlistId);
+        if (m_playlistDao.getHiddenType(playlistId) == PlaylistDAO::PLHT_SET_LOG) {
+            idsToBeUpdated.insert(playlistId);
         }
     }
+    updateChildModel(idsToBeUpdated);
 }
 
 void SetlogFeature::slotPlaylistTableRenamed(int playlistId, const QString& newName) {
     Q_UNUSED(newName);
     //qDebug() << "slotPlaylistTableRenamed() playlistId:" << playlistId;
-    enum PlaylistDAO::HiddenType type = m_playlistDao.getHiddenType(playlistId);
-    if (type == PlaylistDAO::PLHT_SET_LOG ||
-            type == PlaylistDAO::PLHT_UNKNOWN) { // In case of a deleted Playlist
-        clearChildModel();
-        m_lastRightClickedIndex = constructChildModel(playlistId);
-        if (type != PlaylistDAO::PLHT_UNKNOWN) {
-            if (playlistId == m_pPlaylistTableModel->getPlaylist()) {
-                activatePlaylist(playlistId);
-            } else if (m_pSidebarWidget) {
-                m_pSidebarWidget->selectChildIndex(indexFromPlaylistId(playlistId), false);
-            }
-        }
+    if (m_playlistDao.getHiddenType(playlistId) == PlaylistDAO::PLHT_SET_LOG) {
+        updateChildModel(QSet<int>{playlistId});
     }
 }
 
 void SetlogFeature::activate() {
+    // The root item was clicked, so actuvate the current playlist.
+    m_lastClickedIndex = m_pSidebarModel->getRootIndex();
     activatePlaylist(m_playlistId);
 }
 
 void SetlogFeature::activatePlaylist(int playlistId) {
     //qDebug() << "BasePlaylistFeature::activatePlaylist()" << playlistId;
+    if (playlistId == kInvalidPlaylistId) {
+        return;
+    }
     QModelIndex index = indexFromPlaylistId(playlistId);
-    if (playlistId != kInvalidPlaylistId && index.isValid()) {
-        m_pPlaylistTableModel->setTableModel(playlistId);
-        emit showTrackModel(m_pPlaylistTableModel);
-        emit enableCoverArtDisplay(true);
-        // Update selection only, if it is not the current playlist
-        // since we want the root item to be the current playlist as well
-        if (playlistId != m_playlistId) {
-            emit featureSelect(this, index);
-            activateChild(index);
-        }
+    VERIFY_OR_DEBUG_ASSERT(index.isValid()) {
+        return;
+    }
+    emit saveModelState();
+    m_pPlaylistTableModel->setTableModel(playlistId);
+    emit showTrackModel(m_pPlaylistTableModel);
+    emit enableCoverArtDisplay(true);
```

Comment:
why we need this?

Reply:
This is only called by activate above (select the History root). That is not a regular root item because clicking it loads the current playlist (item below the root).

Reply:
Ah OK I understand.